TemperatureSensor: Add Rangefinder to TEMPx_SRC sensor source options

anishk85 opened the pull request:
This PR adds 'Rangefinder' as a valid option to the TEMPx_SRC parameter, allowing an external temperature sensor to override the rangefinder's temperature reading (e.g., for the WATER_DEPTH MAVLink message).

- Adds Rangefinder=9 to AP_TemperatureSensor_Params::Source
- Updates parameter documentation to include 9:Rangefinder
- Implements get_temp_with_override() in RangeFinder
- Uses override in Rover's WATER_DEPTH message
- Enables AP_TemperatureSensor for Rover

This allows users to select an external temperature sensor for water temperature reporting, improving flexibility and accuracy.

Fixes #31748

ES-Alexander commented on libraries/AP_RangeFinder/AP_RangeFinder.cpp:
```diff
@@ -71,6 +71,8 @@
 #include <AP_Vehicle/AP_Vehicle_Type.h>
 #include <AP_HAL/I2CDevice.h>
 #include <AP_InternalError/AP_InternalError.h>
+#include <AP_TemperatureSensor/AP_TemperatureSensor.h>
```
This should be guarded with a precompiler check on whether the temperature sensor library is enabled - otherwise it shouldn’t be included in the build.

magicrub replied:
This is why we have the _config.h files. It is a lightweight way to check/populate the define to see if a library is enabled, then you can check the define to include the main .h

ES-Alexander commented on libraries/AP_RangeFinder/AP_RangeFinder.cpp:
```diff
@@ -857,7 +859,33 @@ bool RangeFinder::get_temp(enum Rotation orientation, float &temp) const
 }
-
+// get temperature (with external sensor override support)
+bool RangeFinder::get_temp_with_override(enum Rotation orientation, float &temp) const
```
I’m unsure on what’s preferable out of renaming the original function to this, or just implementing the override “silently” within the `get_temp` function (since as far as the using code is concerned they’re getting whatever temperature is relevant to the rangefinder).

Either way, I don’t think the internal function (which does not factor in the override) should remain public - that seems like asking for someone to use it without realising they’re supposed to be using the override one, which would lead to bugs where the override is inconsistently applied.

ES-Alexander reviewed (changes requested):
Thanks for the contribution @anishk85!

I’m away for a few days so can’t test at the moment, but had a quick look through the code on my phone. It looks like things are generally in the right direction, but I’ve given some comments to hopefully help flesh out the implementation, and keep the code maintainable for the future :-)

magicrub commented on Rover/GCS_MAVLink_Rover.cpp:
```diff
@@ -228,9 +228,9 @@ void GCS_MAVLINK_Rover::send_water_depth()
+        // get temperature (with external sensor override support)
         float temp_C;
-        if (!s->get_temp(temp_C)) {
+        if (!rangefinder->get_temp_with_override(ROTATION_PITCH_270, temp_C)) {
```
This code should not change. Please look at the other uses of the temp _SRC in other libraries. Look for where AP_TEMPERATURE_SENSOR_ENABLED is used in AP_ESC_Telem and AP_BattMonintor. From the vehicle's perspective it should not know/care where the temperature is coming from, the temp library overrides the rangefinder's value.

If we did it this way, all vehicles would have to check which is not very good abstraction.

ES-Alexander commented on Rover/config.h:
```diff
@@ -55,7 +55,6 @@
 # define MODE_FOLLOW_ENABLED AP_FOLLOW_ENABLED
 #endif
 
-
```
Whitespace should generally not be the only change in a file, especially in an otherwise functional pull request.

ES-Alexander reviewed (changes requested):
Nice improvements - thanks!

I'm still unsure why the `get_temp` and `set_temperature` functions need to be tied to orientation, and doing so without alternatives seems to needlessly disallow multiple rangefinders with the same orientation (which as far as I can tell is a valid and currently viable possibility).

Would it not work to base those methods around instance IDs instead? That seems more direct, less restrictive of the user's options, and likely results in simpler code.

anishk85 commented:
> Nice improvements - thanks!
> 
> I'm still unsure why the `get_temp` and `set_temperature` functions need to be tied to orientation, and doing so without alternatives seems to needlessly disallow multiple rangefinders with the same orientation (which as far as I can tell is a valid and currently viable possibility).
> 
> Would it not work to base those methods around instance IDs instead? That seems more direct, less restrictive of the user's options, and likely results in simpler code.

Hi @ES-Alexander ,
Thanks for reviewing

You’re right — linking get_temp and set_temperature to orientation is too limiting. It also becomes confusing when multiple sensors have the same orientation, like several downward-facing rangefinders.

Using instance IDs is the better solution. It removes this ambiguity by letting us address each sensor directly, regardless of orientation.

  Plan:
   1. Refactor `RangeFinder`: Update get_temp and set_temperature in the
      RangeFinder frontend to accept a sensor instance ID (uint8_t) instead of
      orientation.
   2. Update `AP_TemperatureSensor`: Modify AP_TemperatureSensor_Backend to pass
      the source_id (which is the instance ID) directly to
      RangeFinder::set_temperature.
   3. Update `Rover`: Modify GCS_MAVLink_Rover.cpp to call get_temp on the
      specific backend instance it is iterating over, rather than asking the
      frontend for "any sensor with this orientation."

ES-Alexander reviewed (changes requested):
Generally looking pretty good (to me at least) - nice!

A couple of small remaining issues.
Assuming the ID one gets fixed, I can hopefully test next week :-)

peterbarker commented on libraries/AP_RangeFinder/AP_RangeFinder.cpp:
```diff
@@ -849,15 +850,37 @@ MAV_DISTANCE_SENSOR RangeFinder::get_mav_distance_sensor_type_orient(enum Rotati
+        temp = state[instance].temperature_external;
+        return true;
+    }
+#endif
```
```suggestion
#endif  // AP_TEMPERATURE_SENSOR_ENABLED
```

anishk85 commented:
Hi @ES-Alexander ,
Yeah I changed them locally just asking for the confirmation before pushing those changes

Thanks

anishk85 committed:
TemperatureSensor: Add Rangefinder to TEMPx_SRC sensor source options

Enable Rangefinder as temp source & fix build issues
Uses instance ID for get_temp/set_temperature
Moves override logic to backend to simplify vehicle code
Uses NaN for invalid external temperature

ES-Alexander commented:
The firmware code generally looks ok to me now - I'll try to test on hardware within the next few days, then tag the most relevant maintainers for review.

It would be good if the functionality could be confirmed within an update to [the relevant autotest](https://github.com/ArduPilot/ardupilot/blob/d6d01b62e5e02b8723357d4108cc11c8fd9d6b50/Tools/autotest/rover.py#L6128), using a [simulated temperature sensor](https://ardupilot.org/dev/docs/sitl-i2c-devices.html). That way there's some assurance that future changes won't accidentally break the feature :-)

ES-Alexander reviewed (approved):
Tested and it works as expected - nice!

Was a bit tricky to determine the correct parameter values to use, particularly since temperature sensor values are not automatically reported in MAVLink, and don't have DEV IDs (so I had to check logs to confirm the sensor was actually working/connected). Apparently ArduPilot doesn't directly use MAVLink `DISTANCE_SENSOR` IDs as its instance numbers (which makes some sense), and it's the autopilot's rangefinder instance number that needs to be specified as `TEMPx_SRC_ID`.

ja51d commented:
Hi @anishk85 — saw this PR has been sitting for ~2 months. Are you still planning to address it? Just don't want to
start parallel work if you're still active.

No rush.

anishk85 commented:
Hi @ja51d,
Thanks for the reminder. You can continue working on this since I’m a bit busy right now.
From what I tested, most things are working correctly. Only a few small changes are still needed.

ja51d commented:
Thanks @anishk85 — picking it up. Will open a new PR continuing the work since I don't have push here.

peterbarker commented:
Continued in #33153
